[M2-6231] Add Team Member

farmerpaul opened the pull request:
<!-- Use this template as a guide to describe your pull request, and adjust as necessary. -->
<!-- Include information that helps your peers review your updates and understand this    -->
<!-- repository's history of changes over time.                                           -->

<!-- Delete any options that are not relevant -->

- [x] Tests for the changes have been added

### 📝 Description

<!-- Contributions are welcome! If there is a corresponding      -->
<!-- JIRA ticket, link to it by replacing `#` with ticket number -->

🔗 [Jira Ticket M2-6231](https://mindlogger.atlassian.net/browse/M2-6231)

Created the Add Team Member popup, which is accessed from **Applet > Team > Add Team Member**.

Related changes include:
- Added a `limitTagRows` prop to the `TagsAutocompleteController` component to assign a maximum scrollable height
- Updated input helper/error text font size to align with Figma, and allowed it to word-wrap if needed rather than be truncated with "…" ([Figma convo](https://www.figma.com/file/6BNx63peNVzrHAF1FoZ78x?type=design&node-id=2185-51645&mode=design#773960311) for context)
- Simplified styling of table header on Participants tab to better align with structure used elsewhere

> [!TIP]
> Other minor changes are explained in commit messages, so review [individual commits](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1687/commits) for those descriptions.

### 📸 Screenshots


https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/62e813a0-1bd2-410c-b16f-675c8b0065fe


### 🪤 Peer Testing

Please see [JIRA task's AC](https://mindlogger.atlassian.net/browse/M2-6231) for detailed expectations while testing.

> [!NOTE]
> Currently, team members that have been invited but have not yet accepted the invitation are **not** listed as "pending" on the Team tab, and will be covered in a separate task TBD.

Prerequisites:
- Have an existing applet created in your account that has several participants (10+).
- Have **no managers (team members)** added to your workspace yet

1. Navigate to that applet's **Team** tab, and press the new **Add Team Member** button
    **Expected outcome:** The Add Team Member popup should appear, with the Workspace Name field present and prepopulated with your account name.
2. Fill out fields, including leaving some empty or providing an invalid email address format.
    **Expected outcome:** Validation errors should appear accordingly, and submit button is disabled until form is valid.
3. Fill form out to add a new team member with **Manager** role, and submit.
    **Expected outcome:** A success banner should appear, however the success message will show **undefined undefined** for the first and last name due to a BE bug, which will be [addressed here](https://mindlogger.atlassian.net/browse/M2-6159).
4. Accept the invitation for the new manager you've invited, and refresh the Team page.
    **Expected outcome:** The new team member should be listed.
5. Start adding another team member
    **Expected outcome:** The Add Team Member popup should appear, with the Workspace Name field absent.
6. This time choose **Reviewer** role and select 10+ participants in the Participants field.
    **Expected outcome:** Tags area of tag selector should be capped + scrollable after 3 rows of tags.
7. Fill form out the rest of the form, and submit.
    **Expected outcome:** A success banner should appear, however the success message will again show **undefined undefined** due to the [same BE bug](https://mindlogger.atlassian.net/browse/M2-6159).
8. Accept the invitation for the new reviewer you've invited, and refresh the Team page.
    **Expected outcome:** The new reviewer should be listed (except their Title field, if provided, will not be populated due to the [same BE bug](https://mindlogger.atlassian.net/browse/M2-6159)).

farmerpaul committed:
M2-6231: align input helper/error text w/ designs

Context: https://www.figma.com/file/6BNx63peNVzrHAF1FoZ78x?type=design&node-id=2185-51645&mode=design#773960311
Update font size of form field helper/error text to 14px, and add
support for word-wrapping of longer validation error messages in a way
that doesn't affect layout of existing single-line error messages.

farmerpaul committed:
M2-6231: update types, move to shared location

Support title field and make certain props optional in invitation
request. Move WorkspaceInfo to shared spot for use by new popup.

farmerpaul committed:
M2-6231: update TagsAutocompleteController styling

Add support for limiting # of rows of tags shown. If tags have been
selected and a row limit has been provided, hide input until focused
to better align with designs.

farmerpaul committed:
M2-6231: update Modal styling

Default submit button variant should now be "contained" so it doesn't
need to be specified explicitly in existing popups. Align default Modal
margins to Figma, and update `TakeNowModal` to use common
`StyledModalWrapper` to reuse predefined content padding.

farmerpaul committed:
M2-6231: add AddManagerPopup based on AddUserForm

farmerpaul committed:
M2-6231: remove obsolete BE error case

In studying the BE for AddManagerPopup, learned that the error message
referenced by AddParticipantPopup – EMAIL_IN_USE – is never produced by
the BE (that's what RESPONDENT_ALREADY_INVITED handles), so removed
that case and its associated code.

farmerpaul committed:
M2-6231: integrate AddManagerPopup into Team tab

Also update Participants table header to better match simpler styling
structure already used on similar screens.

farmerpaul committed:
M2-6231: update language strings

Made `noPermissions` message more generic so that it can be used for
content that isn't specifically a "tab". Updated test cases to match.

farmerpaul committed:
M2-6231: ✅ fix modal submit button variant tests

sultanofcardio commented on src/modules/Dashboard/features/Managers/Popups/AddManagerPopup/AddManagerPopup.tsx:
```diff
@@ -0,0 +1,227 @@
+  });
+  const { execute: executeGetWorkspaceInfoApi } = useAsync(getWorkspaceInfoApi, (res) => {
+    setWorkspaceInfo(res?.data?.result || null);
+  });
```
There's a slight delay in showing the workspace name field while this request is resolving. It might make sense to hoist this to the parent and pass the result as a prop. That way the field will show up instantly, and the request won't happen every time the modal opens

https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/343b01bb-2df7-4b3d-a445-3c113cecfd69

farmerpaul replied:
Makes sense, yeah that was bugging me a bit too!

sultanofcardio commented on src/modules/Dashboard/features/Managers/Popups/AddManagerPopup/AddManagerPopup.tsx:
```diff
@@ -0,0 +1,227 @@
+
+    executeGetWorkspaceInfoApi({ ownerId });
+    dispatch(
+      getAllWorkspaceRespondents({
+        params: { ownerId, appletId },
+      }),
```
I don't like the approach of loading "all" the respondents into the dropdown list, but I understand it's the existing pattern. Nothing for you to do here, but I just thought I'd mention it.

I didn't realise this was the approach being used when I implemented the take now modal and ended up doing a debounced search based on user input

farmerpaul replied:
That's a good observation… I'm not able to tell immediately if lazy-load is easy enough to implement here. Both `LabeledUserDropdown` and `TagsAutocompleteController` use the same `Autocomplete` component… lemme see what I can do.

sultanofcardio replied:
No I think we can resolve this without a fix. I just wanted to make a note here

sultanofcardio commented on src/modules/Dashboard/features/Managers/Popups/AddManagerPopup/AddManagerPopup.tsx:
```diff
@@ -0,0 +1,227 @@
+    isLoading,
+  } = useAsync(postAppletInvitationApi, async (result) => {
+    ownerId && executeGetWorkspaceInfoApi({ ownerId });
+    const { firstName, lastName, title } = result.data?.result ?? {};
+    dispatch(
+      banners.actions.addBanner({
+        key: 'AddParticipantSuccessBanner',
+        bannerProps: {
+          id: `${firstName} ${lastName}${title ? `, ${title}` : ''}`,
+        },
+      }),
+    );
```
I've read the description and comments on [M2-6159](https://mindlogger.atlassian.net/browse/M2-6159) and agree that it would be ideal for these fields to be coming back in the response. However, I don't think we should leave the UX like this in the interim. At this point we have a successful response as well as the values that were submitted. I think it's reasonable to assume the values were accepted and display them to the user in the banner. If there is an API validation error, it should return an error response

farmerpaul replied:
Last time I tried to work around this issue by injecting the form values into the banner, they still didn't come through, and I assumed it was because the popup was getting destroyed too soon (in the context of `AddParticipantPopup`), but it doesn't look like that should be happening here. So you're right, I _should_ be able to get the values straight from the form without issue – I'll give it a try.

farmerpaul replied:
Workaround no longer required, fixed in https://github.com/ChildMindInstitute/mindlogger-backend-refactor/pull/1273!

sultanofcardio commented on src/modules/Dashboard/features/Managers/Managers.tsx:
```diff
@@ -129,6 +135,17 @@ export const Managers = () => {
+            ),
+            value: '',
+            width: '8rem',
+          },
```
We [recently decided](https://metalab.slack.com/archives/C06EKU790TY/p1714077416036969?thread_ts=1714075766.398369&cid=C06EKU790TY) to remove these checkboxes for now. See #1681

sultanofcardio reviewed (commented):
@farmerpaul I'm not all the way through yet, but I thought I'd leave my thoughts so far. Hopefully they're not out of date 😅

ChaconC commented on src/modules/Dashboard/features/Managers/Popups/AddManagerPopup/AddManagerPopup.tsx:
```diff
@@ -0,0 +1,227 @@
+    setWorkspaceInfo(res?.data?.result || null);
+  });
+
+  const handleSubmitForm = (values: AddManagerFormValues) => {
```
I can see the table headers being shifted and that's out of scope here, however the manager and coordinator type accounts have a `title` ("Mng Demo" or "Mr") the reviewer type doesn't, is this expected? I filled all the fields for all 4 accounts.

![CleanShot 2024-04-30 at 11 30 48](https://github.com/ChildMindInstitute/mindlogger-admin/assets/2916156/da1c4033-8cbb-4107-b6c3-b37aa81e5d8b)

farmerpaul replied:
I think the columns should be correct now, since I removed the checkbox column?

As for the title, yes, reviewers weren't having their title saved due to the BE bug mentioned in the description, however I just [merged a fix](https://github.com/ChildMindInstitute/mindlogger-backend-refactor/pull/1273) for it, and `title` should be getting accepted now (if you fetch the latest for the BE).

sultanofcardio commented on src/modules/Dashboard/features/Managers/Popups/AddManagerPopup/AddManagerForm/AddManagerForm.tsx:
```diff
@@ -0,0 +1,131 @@
+            name={Fields.email}
+            label={t('emailAddress')}
+            data-testid={`${dataTestId}-email`}
+          />
```
These form fields have placeholders in the designs. Should we include them here as well?

farmerpaul replied:
I'll check with design; the other form was implemented the same and didn't get any feedback about it, so probably good to confirm.

farmerpaul replied:
[Confirmed](https://metalab.slack.com/archives/CH3L0LJR0/p1714516035714119?thread_ts=1714513836.846349&cid=CH3L0LJR0) and fixed!

ChaconC commented on src/modules/Dashboard/features/Participants/Participants.tsx:
```diff
@@ -408,8 +413,9 @@ export const Participants = () => {
-      <ParticipantsHeader>
-        <HeaderSectionLeft>
+
+      <StyledFlexWrap sx={{ gap: 1.2, mb: 2.4 }}>
```
margins should use `theme.spacing`

farmerpaul replied:
They are. 🙂 That's a [built-in shorthand](https://github.com/ChildMindInstitute/mindlogger-backend-refactor/pull/1273) of using the `sx` prop. If I had my druthers we'd be using the shorthand everywhere, but it's only used sporadically so far.

ChaconC commented on src/modules/Dashboard/features/Managers/Popups/AddManagerPopup/AddManagerPopup.tsx:
```diff
@@ -0,0 +1,239 @@
+    defaultValues: defaults,
+    mode: 'all',
+    delayError: VALIDATION_DEBOUNCE_VALUE,
+  });
```
The `delayError` here is causing some errors to not display when changing fields before the delay passes, and you can end up in a form without errors that doesn't let you submit

https://github.com/ChildMindInstitute/mindlogger-admin/assets/2916156/e8bd2a84-3317-4930-a9a2-18d5c1e3abd9


![CleanShot 2024-04-30 at 11 11 31](https://github.com/ChildMindInstitute/mindlogger-admin/assets/2916156/e1eeb4a0-3a8e-4d61-8a01-17fb41f67aaf)

farmerpaul replied:
Good catch! I was able to reproduce this, too. Fixed it and maintained good UX by removing the error delay and running validation `onBlur` instead. Applied same fix to the other two forms I've authored that used the same strategy.

farmerpaul committed:
M2-6231: remove Team table checkbox column

farmerpaul committed:
M2-6231: improve UX by passing workspaceInfo

Have Managers page fetch workspaceInfo ahead of displaying
AddManagerPopup so that there's no flicker of form field showing.

sultanofcardio reviewed (commented):
Still not finished, but it looks like I can't respond to comments while having a pending review

farmerpaul committed:
M2-6231: improve form validation UX

Using onBlur validation without errorDelay avoids the situation of a
field being invalid but its error state not being flagged.

Also remove an unused variable.

farmerpaul committed:
M2-6231: add field placeholders to all user popups

farmerpaul committed:
M2-6231: fix participants field validation
